### PR TITLE
Add support for adding AGS layers to the map

### PR DESCRIPTION
**Pull request by kdeloach**
### Changes

* Add support for AGS layers
* Implement layer filter & clear all
* Implement `excludeLayers` blacklist
* Populate on-demand loaded layers 

Connects #480

### CODE REVIEW

File: src/GeositeFramework/plugins/layer_selector_v2/util.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+define(['underscore'], function(_) {
+    // Differs from the underscore `find` in that this version will return
```
Comment (kdeloach) [inline src/GeositeFramework/plugins/layer_selector_v2/util.js]: TODO strict

File: src/GeositeFramework/plugins/layer_selector_v2/main.js [hunk 2]
```diff
@@ -35,6 +41,71 @@ define([
...
+                if (mapLayer) {
+                    return;
+                }
+
+                // TODO: switch
+                mapLayer = new ArcGISDynamicMapServiceLayer(serviceUrl);
+
+                // Need to assign a deterministic ID, otherwise, the ESRI
+                // JSAPI will generate a unique ID for us.
+                mapLayer.id = serviceUrl;
```
Comment (caseypt) [inline src/GeositeFramework/plugins/layer_selector_v2/main.js]: Does it not use the id defined in the service?

Comment (kdeloach) [inline src/GeositeFramework/plugins/layer_selector_v2/main.js]: Layers have ids but map services don't, as far as I can tell.

**Comment by caseypt** [comment]
This is working well so far. Nice job. I can take another look at it once it's out of WIP.

File: src/GeositeFramework/plugins/layer_selector_v2/main.js (hunk 2 above)
Comment (caseypt) [inline src/GeositeFramework/plugins/layer_selector_v2/main.js]: Makes sense.

**Comment by kdeloach** [comment]
Rebased and updated the PR description.

**Comment by mmcfarland** [comment]
I can't remove a layer by unselecting it (as opposed to unselecting its parent), is that part of the issue you noted in the PR description?

**Comment by kdeloach** [comment]
I think so. I could update this to only allow leaf nodes to be added to the map to make it easier to test.

**Comment by kdeloach** [comment]
Aside from the UI, all outstanding issues have been addressed in the latest commits.

**Comment by caseypt** [comment]
Taking a look now.

**Comment by caseypt** [comment]
When I filter for "land use" and click on the land use item in the layer selector, I get the following error:

screen shot 2016-02-09 at 12 36 51 pm

**Comment by caseypt** [comment]
+1. This is really impressive. Nice job. There is a lot of code here, but it is all very readable. I've tried to focus on finding bugs, but I've only run into the one I reported above.

**Comment by caseypt** [comment]
Full stack trace

image

**Comment by kdeloach** [comment]
I was able to reproduce that error by doing this:

1. Click "Hurricane Sandy"
2. Apply filter ("land use")
3. Click the "Land Use" layer

**Comment by kdeloach** [comment]
Fixed.

**Comment by caseypt** [comment]
+1. Verified the fix locally.

**Comment by kdeloach** [comment]
Thanks.

**Pull request merged**